src/types.rs

```diff
@@ -236,63 +332,73 @@ impl Decode for PageSize {
 }
 
 /// [Memory type](https://webassembly.github.io/spec/core/binary/types.html#memory-types).
-#[cfg_attr(
-    all(not(feature = "threads"), not(feature = "custom-page-sizes")),
-    derive(Wasmbin)
-)]
-#[derive(WasmbinCountable, Debug, PartialEq, Eq, Hash, Clone, Visit)]
+#[derive(Wasmbin, WasmbinCountable, Debug, PartialEq, Eq, Hash, Clone, Visit)]
 pub struct MemType {
-    #[cfg(feature = "custom-page-sizes")]
-    pub page_size: Option<PageSize>,
-    #[cfg(feature = "threads")]
-    pub is_shared: bool,
     pub limits: Limits,
 }
 
-#[cfg(all(feature = "threads", not(feature = "custom-page-sizes")))]
-encode_decode_as!(MemType, {
-    (MemType { is_shared: false, limits: Limits { min, max: None } }) <=> (MemTypeRepr::Unshared(LimitsRepr::Min { min })),
-    (MemType { is_shared: false, limits: Limits { min, max: Some(max) } }) <=> (MemTypeRepr::Unshared(LimitsRepr::MinMax { min, max })),
-    (MemType { is_shared: true, limits: Limits { min, max: None } }) <=> (MemTypeRepr::SharedMin { min }),
-    (MemType { is_shared: true, limits: Limits { min, max: Some(max) } }) <=> (MemTypeRepr::SharedMinMax { min, max }),
-});
+/// [Reference type](https://webassembly.github.io/spec/core/binary/types.html#reference-types).
+#[derive(Wasmbin, Debug, PartialEq, Eq, Hash, Clone, Visit)]
+#[repr(u8)]
+pub enum RefType {
```

Comment:
Wow, the hierarchy here is getting pretty complicated.

I guess not much way around it if that's how spec structures these types, but it seems some things are inherent to our structures?

Eg I noticed that HeapType has AbstractHeapType as one of the variants, but then in this enum we support both HeapType and AbstractHeapType at higher level... is that a bug?

Are RefType::HeapType::Abstract and RefType::AbstractHeapType really used to represent different things?